lib/puppet/provider/ec2_volume/v2.rb

```diff
@@ -1,31 +1,40 @@
 require_relative '../../../puppet_x/puppetlabs/aws'
 
-Puppet::Type.type(:ec2_volume).provide(:v2, :parent => PuppetX::Puppetlabs::Aws) do
+Puppet::Type.type(:ec2_volume).provide(:v2, parent: PuppetX::Puppetlabs::Aws) do
   confine feature: :aws
+  confine feature: :retries
 
   mk_resource_methods
 
+  @@RETRIES = 10
+
   def self.instances
     regions.collect do |region|
       ec2 = ec2_client(region)
       begin
         volumes = []
-        volume_response = ec2.describe_volumes()
+        volume_response = ec2.describe_volumes
         volume_response.data.volumes.collect do |volume|
           hash = volume_to_hash(region, volume)
           volumes << new(hash) if has_name?(hash)
         end
         volumes
       rescue Timeout::Error, StandardError => e
-        raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, self.resource_type.name.to_s, e.message)
+        raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, resource_type.name.to_s, e.message)
       end
     end.flatten
   end
 
   def self.prefetch(resources)
-    instances.each do |prov|
-      if resource = resources[prov.name] # rubocop:disable Lint/AssignmentInCondition
-        resource.provider = prov if resource[:region] == prov.region
+    with_retries(:max_tries => @@RETRIES,
+                 :rescue => Aws::EC2::Errors::RequestLimitExceeded,
+                 :base_sleep_seconds => 30,
+                 :max_sleep_seconds => 60) do |attempt|
+      Puppet.notice("Attempt #{attempt} of prefetch ebs volume")
+      instances.each do |prov|
```

Comment:
I *really* think we should stop describing all volumes, just to loop over them to find the ones we're interested in. Calling describe-volumes (with no filter) on our account takes roughly 6.5s...
```
$ time aws ec2 describe-volumes 1>/dev/null

real	0m6.436s
user	0m1.088s
sys	0m0.131s
```
Calling it with a filter is about 10x faster...
```
$ time aws ec2 describe-volumes --filters "Name=tag:Name,Values=internal.msm.gb.perf1.clubs-mse.clubs-mongodb-wt3.3.data" 1>/dev/null

real	0m0.690s
user	0m0.410s
sys	0m0.118s
```